### PR TITLE
TokenNetwork - 2 nodes can have maximum 1 channel

### PR DESCRIPTION
**Pull request by loredanacirstea**
fixes https://github.com/raiden-network/raiden-contracts/issues/40

Structural changes for limiting the number of channels between two nodes to 1.
All changes stem from changing the `channel_identifier` from a monotonically increasing `uint256` to the `keccak256` of the two channel participant addresses, ordered lexicographically.

Changes:

- `channel_identifier` from `uint256` to `bytes32` - for all events
- `openChannel` returns `bytes32`
- `setDeposit(address participant, uint256 total_deposit, address partner)`
- getters:
  - `function getChannelIdentifier(address participant, address partner)` returns `bytes32`
  - `function getChannelInfo(address participant1, address participant2)` returns `(channel_identifier, settle_block_number, state)`
  - `function getChannelParticipantInfo(address participant, address partner)` returns `(deposit, is_the_closer, balance_hash, nonce)`
  - `function getParticipantLockedAmount(address participant1, address participant2, bytes32 locksroot)` returns `locked_amount`
- other abi changes:
```
function closeChannel(
        address partner,
        bytes32 balance_hash,
        uint256 nonce,
        bytes32 additional_hash,
        bytes signature
    )
```
```
function updateNonClosingBalanceProof(
        address closing_participant,
        address non_closing_participant,
        bytes32 balance_hash,
        uint256 nonce,
        bytes32 additional_hash,
        bytes closing_signature,
        bytes non_closing_signature
    )
```
```
function settleChannel(
        address participant1,
        uint256 participant1_transferred_amount,
        uint256 participant1_locked_amount,
        bytes32 participant1_locksroot,
        address participant2,
        uint256 participant2_transferred_amount,
        uint256 participant2_locked_amount,
        bytes32 participant2_locksroot
    )
```
```
function unlock(
        address participant,
        address partner,
        bytes merkle_tree_leaves
    )
```
```
function cooperativeSettle(
        address participant1_address,
        uint256 participant1_balance,
        address participant2_address,
        uint256 participant2_balance,
        bytes participant1_signature,
        bytes participant2_signature
    )
```

## New gas costs 

```
TokenNetworksRegistry DEPLOYMENT 3836879
TokenNetworksRegistry.createERC20TokenNetwork 2547850
SecretRegistry.registerSecret 45414
TokenNetwork.openChannel 67891
TokenNetwork.setDeposit 54760
TokenNetwork.closeChannel 109590
TokenNetwork.updateNonClosingBalanceProof 91011
TokenNetwork.settleChannel 78445
TokenNetwork.unlock 1 locks 40209
TokenNetwork.unlock 6 locks 74128
```

### CODE REVIEW

**Comment by LefterisJP** [comment]
Reviewing this PR.

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 1]
```diff
@@ -647,58 +633,60 @@ contract TokenNetwork is Utils {
...
         pure
-        internal
-        returns (bytes32 data)
+        public
+        returns (bytes32)
     {
-        return keccak256(channel_identifier, participant, partner);
+        // Lexicographic order of the channel addresses
+        // This limits the number of channels that can be opened between two nodes to 1.
+        if (participant < partner) {
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: Why not move the check for `participant == partner` in here and revert if they are the same?

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 2]
```diff
@@ -224,27 +210,33 @@ contract TokenNetwork is Utils {
...
-        isOpen(channel_identifier)
-        isParticipant(channel_identifier, msg.sender)
+        isOpen(msg.sender, partner)
         public
     {
-        address partner_address;
+        require(partner != 0x0);
+
+        // Message sender and partner must be different
+        require(msg.sender != partner);
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: I would move this in `getChannelIdentifier()`

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 3]
```diff
@@ -300,6 +289,17 @@ contract TokenNetwork is Utils {
     )
         external
     {
+        require(closing_participant != 0x0);
+        require(non_closing_participant != 0x0);
+        require(closing_participant != non_closing_participant);
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: I would move this check in `getChannelIdentifier()`

File: raiden_contracts/contracts/TokenNetwork.sol [hunk 4]
```diff
@@ -519,49 +497,55 @@ contract TokenNetwork is Utils {
...
         address partner,
         bytes merkle_tree_leaves
     )
         public
     {
-        // Channel must be settled and channel data deleted
-        require(channels[channel_identifier].state == 0);
+        require(participant != 0x0);
+        require(partner != 0x0);
+        require(merkle_tree_leaves.length > 0);
```
Comment (LefterisJP) [inline raiden_contracts/contracts/TokenNetwork.sol]: For example here you forgot to also add the `participant != partner` check. It's much better to put the logic inside the `getChannelIdentifier()` so you don't have to remember it in each place you want to get the channel Identifier.

Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: Yes, doing that. good idea.

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 1 above)
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: I could also move all the `require(participant != 0x0);` checks there

File: raiden_contracts/contracts/TokenNetwork.sol (hunk 4 above)
Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: I could also move all the `require(participant != 0x0);` checks there

Comment (LefterisJP) [reply raiden_contracts/contracts/TokenNetwork.sol]: yeah that's correct! Do you only do those checks at the same places where this `getChannelIdentifier()` gets called? If yes it's perfect. If not you can still include it in the function but I am afraid we may forget the call is in there and end up (in the future) having double checks.

But in general should be fine to include that check too as long as we remember it's in the function.

Comment (loredanacirstea) [reply raiden_contracts/contracts/TokenNetwork.sol]: >  Do you only do those checks at the same places where this getChannelIdentifier() gets called?

Yes. And I moved them. Better have them doubled than not have them at all anyway.

**Comment by LefterisJP** [comment]
I am satisfied by the changes. PR approved.

**Pull request merged**